packages/salesforcedx-vscode-core/src/deploys/toolingDeploy.ts

```diff
@@ -0,0 +1,190 @@
+    const apiVersion =
+      this.apiVersion ||
+      document.getElementsByTagName('apiVersion')[0].textContent;
+    const statusNode = document.getElementsByTagName('status')[0];
+    const packageNode = document.getElementsByTagName('packageVersions')[0];
+
+    const metadataField = {
+      apiVersion,
+      ...(statusNode ? { status: statusNode.textContent } : {}),
+      ...(packageNode ? { packageVersions: packageNode.textContent } : {})
+    };
+    return metadataField;
+  }
+
+  public async createContainerMember(outboundFiles: string[]) {
+    try {
+      const id = this.container!.id;
+      const metadataContent = fs.readFileSync(outboundFiles[1], 'utf8');
+      const metadataField = this.buildMetadataField(metadataContent);
+      const body = fs.readFileSync(outboundFiles[0], 'utf8');
+      const fileName = path.basename(
+        outboundFiles[0],
+        path.extname(outboundFiles[0])
+      );
+
+      const contentEntity = (await this.connection!.tooling.sobject(
+        this.metadataType!
+      ).find({ Name: fileName }))[0];
+
+      const containerMemberObject = {
+        MetadataContainerId: id,
+        FullName: fileName,
+        Body: body,
+        Metadata: metadataField,
+        ...(contentEntity ? { contentEntityId: contentEntity.Id } : {})
+      };
+      const containerMember = (await this.connection!.tooling.create(
+        supportedToolingTypes.get(this.metadataType!)!,
+        containerMemberObject
+      )) as ToolingCreateResult;
+
+      const deployFailed = new Error();
+      if (!containerMember.success) {
+        deployFailed.message = 'Unexpected error creating apex class member';
+        deployFailed.name = 'ApexClassMemberCreationFailed';
+        throw deployFailed;
+      }
+      this.containerMember = containerMember;
+    } catch (e) {
+      throw e;
+    }
+  }
+
+  public async createContainerAsyncRequest() {
+    try {
+      const contAsyncRequest = (await this.connection!.tooling.create(
+        'ContainerAsyncRequest',
+        { MetadataContainerId: this.container!.id }
+      )) as ToolingCreateResult;
+
+      const deployFailed = new Error();
+      if (!contAsyncRequest.success) {
+        deployFailed.message =
+          'Unexpected error creating container async request';
+        deployFailed.name = 'ContainerAsyncRequestFailed';
+        throw deployFailed;
+      }
+      this.asyncRequest = contAsyncRequest;
+    } catch (e) {
+      throw e;
+    }
+  }
+
+  private sleep(ms: number) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  public async toolingRetrieve(): Promise<ToolingRetrieveResult> {
+    try {
+      let retrieveResult = (await this.connection!.tooling.retrieve(
+        'ContainerAsyncRequest',
+        this.asyncRequest!.id
+      )) as ToolingRetrieveResult;
+      let count = 0;
+      while (retrieveResult.State === 'Queued' && count <= 20) {
+        await this.sleep(100);
```

Comment:
We should take a closer look at why we need this sleep. Maybe we can replace this with a pattern that is more reliable.

Reply:
I think it's due to the fact that the deploy is being done with a ContainerAsyncRequest which requires polling the status

Reply:
Yeah, it's because we're polling the status of the ContainerAsyncRequest to see when it changes from 'Queued' to anything else